modules/substrate/src/verifier.rs

```diff
@@ -0,0 +1,667 @@
+
+/// Errors which can happen while importing a header.
+#[derive(Debug, PartialEq)]
+pub enum ImportError {
+	/// This header is older than our latest finalized block, thus not useful.
+	OldHeader,
+	/// This header has already been imported by the pallet.
+	HeaderAlreadyExists,
+	/// This header has never been imported by the pallet.
+	UnknownHeader,
+	/// We're missing a parent for this header.
+	MissingParent,
+	/// We were unable to prove finality for this header.
+	UnfinalizedHeader,
+	/// We failed to verify this header's ancestry.
+	AncestryCheckFailed,
+	/// The header is missing digests related to consensus events.
+	///
+	/// This will typically have to do with missing authority set change signals.
+	MissingConsensusDigest,
+	/// Trying to prematurely import a justification
+	PrematureJustification,
+	/// The height of the next authority set change overflowed.
+	ScheduledHeightOverflow,
+}
+
+/// A trait for verifying whether a header is valid for a particular blockchain.
+pub trait ChainVerifier<S, H: HeaderT> {
+	/// Import a header to the pallet.
+	fn import_header(storage: &mut S, header: H) -> Result<(), ImportError>;
+
+	/// Verify that the given header has been finalized and is part of the canonical chain.
+	fn verify_finality(storage: &mut S, hash: H::Hash, proof: FinalityProof) -> Result<(), ImportError>;
+}
+
+/// Used to verify imported headers and their finality status.
+#[derive(Debug)]
+pub struct Verifier;
+
+impl<S, H> ChainVerifier<S, H> for Verifier
+where
+	S: BridgeStorage<Header = H>,
+	H: HeaderT,
+{
+	fn import_header(storage: &mut S, header: H) -> Result<(), ImportError> {
+		let best_finalized = storage.best_finalized_header();
+
+		if header.number() < best_finalized.number() {
+			return Err(ImportError::OldHeader);
+		}
+
+		if storage.header_exists(header.hash()) {
+			return Err(ImportError::HeaderAlreadyExists);
+		}
+
+		let parent_header = storage.get_header_by_hash(*header.parent_hash());
+		if parent_header.is_none() {
+			return Err(ImportError::MissingParent);
+		}
+
+		// We don't need the justification right away, but we should note it.
+		let requires_justification = find_scheduled_change(&header).is_some();
+		let is_finalized = false;
+		storage.write_header(&ImportedHeader::new(header, requires_justification, is_finalized));
+
+		Ok(())
+	}
+
+	fn verify_finality(storage: &mut S, hash: H::Hash, proof: FinalityProof) -> Result<(), ImportError> {
+		// Make sure that we've previously imported this header
+		let header = storage.get_header_by_hash(hash).ok_or(ImportError::UnknownHeader)?;
+
+		let current_authority_set = storage.current_authority_set();
+		let is_finalized = prove_finality(&header, &current_authority_set, proof);
+		if !is_finalized {
+			return Err(ImportError::UnfinalizedHeader);
+		}
+
+		let last_finalized = storage.best_finalized_header();
+		let mut finalized_headers = if let Some(ancestors) = are_ancestors(storage, last_finalized, header.clone()) {
```

Comment:
This method only works if `last_finalized.block_number() < header.block_number()` but we don't check it anywhere. Without the check the `are_ancestors` function may traverse the entire blockchain up to genesis before it fails.

Reply:
Proposed a change to `are_ancestors` to prevent that.

Reply:
> Without the check the `are_ancestors` function may traverse the entire blockchain up to genesis before it fails.

Yes, good catch!